pkg/meta/redis_lock.go

```diff
@@ -105,54 +106,47 @@ func dumpLocks(ls []plockRecord) []byte {
 	return wb.Bytes()
 }
 
-func insertLocks(ls []plockRecord, i int, nl plockRecord) []plockRecord {
-	nls := make([]plockRecord, len(ls)+1)
-	copy(nls[:i], ls[:i])
-	nls[i] = nl
-	copy(nls[i+1:], ls[i:])
-	ls = nls
-	return ls
-}
-
 func updateLocks(ls []plockRecord, nl plockRecord) []plockRecord {
 	// ls is ordered by l.start without overlap
-	var i int
-	for i < len(ls) && nl.end > nl.start {
+	for i := 0; i < len(ls) && nl.start <= nl.end; i++ {
 		l := ls[i]
-		if l.end < nl.start {
-		} else if l.start < nl.start {
-			ls = insertLocks(ls, i+1, plockRecord{nl.ltype, nl.pid, nl.start, l.end})
-			ls[i].end = nl.start
-			i++
-			nl.start = l.end
-		} else if l.end < nl.end {
-			ls[i].ltype = nl.ltype
-			ls[i].start = nl.start
-			nl.start = l.end
-		} else if l.start < nl.end {
-			ls = insertLocks(ls, i, nl)
-			ls[i+1].start = nl.end
-			nl.start = nl.end
-		} else {
-			ls = insertLocks(ls, i, nl)
-			nl.start = nl.end
+		if nl.start < l.start && nl.end >= l.start {
+			// split nl
+			ls = append(ls, nl)
+			ls[len(ls)-1].end = l.start - 1
```

Comment:
nl is spliited to two parts: 1. last item of ls; 2: new nl
1 is not handled properly